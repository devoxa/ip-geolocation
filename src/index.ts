import { Reader } from 'mmdb-lib'
import fs from 'fs/promises'
import path from 'path'

export interface GeolocateIpResult {
  continent: { code: string; name: string }
  country: { code: string; name: string; isInEuropeanUnion: boolean }
  subdivision: { name: string }
  city: { name: string }
  location: { latitude: number; longitude: number }
}

const DB_FILE_PATH = path.join(__dirname, 'dbip-city-lite.mmdb')

export async function geolocateIp(ip: string): Promise<GeolocateIpResult | null> {
  const databaseBuffer = await fs.readFile(DB_FILE_PATH)

  const reader = new Reader(databaseBuffer)
  const result = reader.get(ip) as any // eslint-disable-line @typescript-eslint/no-explicit-any

  if (!result) {
    return null
  }

  return {
    continent: {
      code: result.continent.code,
      name: result.continent.names.en,
    },
    country: {
      code: result.country.iso_code,
      name: result.country.names.en,
      isInEuropeanUnion: result.country.is_in_european_union,
    },
    subdivision: {
      name: result.subdivisions[0].names.en,
    },
    city: {
      name: result.city.names.en,
    },
    location: {
      latitude: result.location.latitude,
      longitude: result.location.longitude,
    },
  }
}
