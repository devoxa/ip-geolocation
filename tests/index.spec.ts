import { geolocateIp, loadDatabase } from '../src/index'

const PROPERTY_MATCHER = {
  city: {
    name: expect.any(String),
  },
  location: {
    latitude: expect.any(Number),
    longitude: expect.any(Number),
  },
}

describe('ip-geolocation', () => {
  it('errors if the database does not exist', async () => {
    let error

    try {
      await loadDatabase('foo.mmdb')
    } catch (err) {
      error = err
    }

    expect(error).toMatchSnapshot()
  })

  it('can geolocate ip address from the USA', async () => {
    expect(await geolocateIp('69.10.63.243')).toMatchSnapshot(PROPERTY_MATCHER)
  })

  it('can geolocate ip address from the UK', async () => {
    expect(await geolocateIp('217.138.196.20')).toMatchSnapshot(PROPERTY_MATCHER)
  })

  it('can geolocate ip address from Germany', async () => {
    expect(await geolocateIp('78.159.96.195')).toMatchSnapshot(PROPERTY_MATCHER)
  })

  it('can geolocate ip address from South Africa', async () => {
    expect(await geolocateIp('172.107.93.212')).toMatchSnapshot(PROPERTY_MATCHER)
  })

  it('fails gracefully for invalid or unresolvable ip addresses', async () => {
    expect(await geolocateIp('127.0.0.1')).toEqual(null)
    expect(await geolocateIp('foobar')).toEqual(null)
    expect(await geolocateIp('999.138.196.20')).toEqual(null)
    expect(await geolocateIp('')).toEqual(null)
  })
})
