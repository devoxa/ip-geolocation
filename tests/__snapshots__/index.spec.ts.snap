// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ip-geolocation can geolocate ip address from Germany 1`] = `
Object {
  "city": Object {
    "name": "Frankfurt am Main (Innenstadt I)",
  },
  "continent": Object {
    "code": "EU",
    "name": "Europe",
  },
  "country": Object {
    "code": "DE",
    "isInEuropeanUnion": true,
    "name": "Germany",
  },
  "location": Object {
    "latitude": 50,
    "longitude": 9,
  },
  "subdivision": Object {
    "name": "Hesse",
  },
}
`;

exports[`ip-geolocation can geolocate ip address from South Africa 1`] = `
Object {
  "city": Object {
    "name": "Johannesburg",
  },
  "continent": Object {
    "code": "AF",
    "name": "Africa",
  },
  "country": Object {
    "code": "ZA",
    "isInEuropeanUnion": false,
    "name": "South Africa",
  },
  "location": Object {
    "latitude": -26,
    "longitude": 28,
  },
  "subdivision": Object {
    "name": "Gauteng",
  },
}
`;

exports[`ip-geolocation can geolocate ip address from the UK 1`] = `
Object {
  "city": Object {
    "name": "Scholes",
  },
  "continent": Object {
    "code": "EU",
    "name": "Europe",
  },
  "country": Object {
    "code": "GB",
    "isInEuropeanUnion": false,
    "name": "United Kingdom",
  },
  "location": Object {
    "latitude": 54,
    "longitude": -1,
  },
  "subdivision": Object {
    "name": "England",
  },
}
`;

exports[`ip-geolocation can geolocate ip address from the USA 1`] = `
Object {
  "city": Object {
    "name": "Secaucus",
  },
  "continent": Object {
    "code": "NA",
    "name": "North America",
  },
  "country": Object {
    "code": "US",
    "isInEuropeanUnion": false,
    "name": "United States",
  },
  "location": Object {
    "latitude": 41,
    "longitude": -74,
  },
  "subdivision": Object {
    "name": "New Jersey",
  },
}
`;

exports[`ip-geolocation errors if the database does not exist 1`] = `[Error: Database file at foo.mmdb does not exist]`;
